lib/puppet/util/windows/taskscheduler2.rb

```diff
@@ -0,0 +1,707 @@
+      definition.Principal.UserId = 'SYSTEM'
+      definition.Principal.LogonType = TASK_LOGON_SERVICE_ACCOUNT
+      definition.Principal.RunLevel = TASK_RUNLEVEL_HIGHEST
+      return true
+    else
+      # TODO!!!
+      raise NotImplementedError
+    end
+  end
+
+  # Returns the user associated with the task or nil if no user has yet
+  # been associated with the task.
+  # DONE
+  def principal
+    raise Error.new(_('No currently active task. ITask is NULL.')) if definition.nil?
+
+    definition.Principal
+  end
+
+  # Returns the compatibility level of the task.
+  #
+  # DONE
+  def compatibility
+    raise Error.new(_('No currently active task. ITask is NULL.')) if definition.nil?
+
+    definition.Settings.Compatibility
+  end
+
+  # Sets the compatibility with the task.
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa381846(v=vs.85).aspx
+  #
+  def compatibility=(value)
+    # TODO Do we need warnings about this?  could be dangerous?
+    definition.Settings.Compatibility = value
+  end
+
+  # Returns the task's priority level. Possible values are 'idle',
+  # 'normal', 'high', 'realtime', 'below_normal', 'above_normal',
+  # and 'unknown'.
+  # Note - This is an approximation due to how the priority class and thread priority
+  # levels differ
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa383070(v=vs.85).aspx
+  #
+  # DONE
+  def priority
+    case priority_value
+    when 0
+      'realtime'
+    when 1
+      'high'
+    when 2,3
+      'above_normal'
+    when 4,5,6
+      'normal'
+    when 7,8
+      'below_normal'
+    when 9,10
+      'idle'
+    else
+      'unknown'
+    end
+  end
+
+  # Returns the task's priority level as an integer
+  #
+  # DONE
+  def priority_value
+    raise Error.new(_('No currently active task. ITask is NULL.')) if definition.nil?
+    
+    definition.Settings.Priority
+  end
+
+  # Sets the priority of the task. The +priority+ should be a numeric
+  # priority constant value, from 0 to 10 inclusive
+  #
+  # DONE
+  def priority_value=(value)
+    raise Error.new(_('No currently active task. ITask is NULL.')) if definition.nil?
+    raise TypeError unless value.is_a?(Numeric)
+    raise TypeError if value < 0
+    raise TypeError if value > 10
+
+    definition.Settings.Priority = value
+
+    value
+  end
+
+  def new_task_defintion(task_name)
+    raise Error.new(_("task '%{task}' already exists") % { task: task_name }) if exists?(task_name)
+
+    @pITaskDefinition = @task_service.NewTask(0)
+    @task_defn_path = task_name
+    @pITask = nil
+
+    true
+  end
+
+  # Returns the number of actions associated with the active task.
+  #
+  # DONE
+  def action_count
+    raise Error.new(_('No currently active task. ITask is NULL.')) if definition.nil?
+
+    definition.Actions.count
+  end
+
+  # DONE
+  def action(index)
+    raise Error.new(_('No currently active task. ITask is NULL.')) if definition.nil?
+
+    action = nil
+
+    begin
```

Comment:
Need to look at behavior of `puppet resource` to see how these come out?

May need to hide this for the moment as we build out compatibility layer?  Not sure just yet.. need to do some more homework.

Reply:
Unlike `trigger`, we're returning a COM instance of an Action object - do we want that, given this is a leaky interface?